rir/src/compiler/opt/gvn.cpp

```diff
@@ -0,0 +1,204 @@
+                            klassNumber++;
+                        storeNumber(i, klassNumber, args);
+                    } else {
+                        number[i] = klassNumber;
+                    }
+                    changed = true;
+                }
+            };
+
+            if (auto ld = LdConst::Cast(i)) {
+                SEXP constant = ld->c();
+                for (auto& c : constants) {
+                    if (R_compute_identical(c.first, constant, 0)) {
+                        number[i] = c.second;
+                        changed = true;
+                        return;
+                    }
+                }
+                while (classes.count(nextNumber))
+                    nextNumber++;
+                constants.emplace(constant, nextNumber);
+                storeNumber(i, nextNumber, {nextNumber});
+            } else {
+                // We allow instructions with those effects to be deduplicated.
+                auto effects =
+                    i->effects & (~(Effects(Effect::Error) | Effect::Warn |
+                                    Effect::Visibility | Effect::Force));
+                if (effects.empty())
+                    computeNumber(i);
+                else
+                    assignNumber(i);
+            }
+        };
+
+        while (changed) {
+            changed = false;
+            DominatorTreeVisitor<>(dom).run(cls->entry, [&](BB* bb) {
+                for (auto i : *bb)
+                    computeGN(i);
+            });
+        }
+
+        for (auto& g : number) {
+            reverseNumber[g.second].insert(g.first);
+        }
+
+        for (auto it = reverseNumber.begin(); it != reverseNumber.end();) {
+            if (it->second.size() < 2)
+                it = reverseNumber.erase(it);
+            else
+                it++;
+        }
+        if (reverseNumber.size() == 0)
+            return;
+    }
+
+    {
+        std::unordered_map<Value*, Value*> replacements;
+
+        for (auto& g : reverseNumber) {
+            auto p = g.second.begin();
+            auto first = firstValue.at(g.first);
+            while (replacements.count(first))
+                first = replacements.at(first);
+            auto firstInstr = Instruction::Cast(first);
+
+            while (p != g.second.end()) {
+                auto i = Instruction::Cast(*p);
+                p++;
+                if (i && i != first) {
+                    if (firstInstr) {
+                        if (i->bb() == firstInstr->bb()) {
+                            if (!i->bb()->before(firstInstr, i))
+                                continue;
+                        } else {
+                            if (!dom.dominates(firstInstr->bb(), i->bb()))
+                                continue;
+                        }
+                    }
+                    i->replaceUsesWith(first);
+                    // Make sure this instruction really gets removed
+                    i->effects.reset();
```

Comment:
Why this? Because of the error, etc effects?

Reply:
the idea is, that we ignore error/warning for gvn. the reasoning is that the error/warning would already have happened, in the instruction we use to replace with. but then the intruction we replaced, won't be removed, since it is not effectless. so I clear the effects here, such that cleanup pass can remove it